{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "import math\n",
    "import numpy as np\n",
    "import numpy.matlib\n",
    "import random\n",
    "import os\n",
    "import pathlib\n",
    "import time"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Classes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Individual:\n",
    "  def __init__(self,c, w, fitness=math.inf):\n",
    "    self.c = c\n",
    "    self.w = w\n",
    "    self.fitness = fitness\n",
    "\n",
    "  # Getters\n",
    "  def getFitness(self):\n",
    "    return self.fitness\n",
    "\n",
    "  # Setters\n",
    "  def setFitness(self,fitness):\n",
    "    self.fitness = fitness\n",
    "  \n",
    "  # Methods\n",
    "  def printGenome(self):\n",
    "    return print(\"C (size): {0}, W (size): {1}\\nC: {2}\\nW: {3}\\n fitness: {4}\".format(self.c.shape, self.w.shape, self.c, self.w, self.fitness))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Population:\n",
    "  def __init__(self,size):\n",
    "    self.size = size\n",
    "    self.geneList = []\n",
    "    self.fitnessList = []\n",
    "    self.bestGene = None\n",
    "\n",
    "  # Methods\n",
    "  def addGeneToList(self,gen):\n",
    "    self.geneList.append(gen);\n",
    "\n",
    "  def addFitnessToList(self,fitness):\n",
    "    self.fitnessList.append(fitness);\n",
    "\n",
    "  def insertBestGene(self,gen):\n",
    "    self.bestGene = gen\n",
    "    self.geneList.append(gen)\n",
    "    self.fitnessList.append(gen.fitness)\n",
    "\n",
    "  def initializePopulation(self):\n",
    "    for i in range(self.size):\n",
    "      gen = Individual(random_C_generator(K), random_W_generator(K,optimalD))\n",
    "      self.geneList.append(gen)\n",
    "    \n",
    "    # Once the population is set, we choose a random best gene\n",
    "    randomIndex = np.random.randint(low=0, high=self.size)\n",
    "    self.bestGene = self.geneList[randomIndex]\n",
    "\n",
    "    return self"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Particle(Individual):\n",
    "  def __init__(self,c, w, fitness=math.inf):\n",
    "    super().__init__(c, w, fitness)\n",
    "    self.bestFitness = fitness\n",
    "    self.wPosition = w\n",
    "    self.cPosition = c\n",
    "    self.wVelocity = 0 # All swarm's velocities are set to 0 at the start\n",
    "    self.cVelocity = 0 \n",
    "\n",
    "  # Setters\n",
    "  def setNewWeightsPosition(self, new_W):\n",
    "    self.wPosition = new_W\n",
    "    \n",
    "  def setNewChromosomePosition(self, new_C):\n",
    "    self.cPosition = new_C\n",
    "\n",
    "  def setNewWeightsVelocity(self, new_W):\n",
    "    self.wVelocity = new_W\n",
    "\n",
    "  def setNewChromosomeVelocity(self, new_C):\n",
    "    self.cVelocity = new_C\n",
    "\n",
    "  def setBestFitness(self,newF):\n",
    "    self.bestFitness = newF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Swarm(Population):\n",
    "  def __init__(self,size):\n",
    "    super().__init__(size)\n",
    "    self.globalBestW = None\n",
    "    self.globalBestC = None\n",
    "\n",
    "  # Getters\n",
    "  def getBestSwarmWeights(self):\n",
    "    return self.globalBestW\n",
    "\n",
    "  def getBestSwarmChromosome(self):\n",
    "    return self.globalBestC\n",
    "\n",
    "  # Setters\n",
    "  def setBestSwarmWeigths(self,newW):\n",
    "    self.globalBestW = newW\n",
    "\n",
    "  def setBestSwarmChromosome(self,newC):\n",
    "    self.globalBestC = newC\n",
    "\n",
    "  def initializeSwarm(self):\n",
    "    for i in range(self.size):\n",
    "      p = Particle(random_C_generator(K), random_W_generator(K,optimalD))\n",
    "      self.geneList.append(p)\n",
    "          \n",
    "      # Once the swarm is set, we choose a random best gene\n",
    "    randomIndex = np.random.randint(low=0, high=self.size)\n",
    "    self.bestGene = self.geneList[randomIndex]\n",
    "    self.globalBestC = self.bestGene.c\n",
    "    self.globalBestW = self.bestGene.w\n",
    "\n",
    "    return self"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "class Reef:\n",
    "  def __init__(self,rho0, R):\n",
    "    self.reefSize = R\n",
    "    self.freeOccupiedRate = rho0\n",
    "    self.geneList = np.full([R,1], None)\n",
    "    self.fitnessList = []\n",
    "    self.bestGene = None\n",
    "\n",
    "  # Methods\n",
    "  def addGeneToList(self,gen):\n",
    "    self.geneList.append(gen);\n",
    "\n",
    "  def addFitnessToList(self,fitness):\n",
    "    self.fitnessList.append(fitness);\n",
    "\n",
    "  def insertBestGene(self,gen):\n",
    "    self.bestGene = gen\n",
    "    self.geneList.append(gen)\n",
    "    self.fitnessList.append(gen.fitness)\n",
    "\n",
    "  def initializeReef(self):\n",
    "    occupiedHoles = int(self.reefSize * self.freeOccupiedRate)# Partially Occupied\n",
    "    print(occupiedHoles)\n",
    "    for i in range(occupiedHoles): \n",
    "      # Choose random hole\n",
    "      random_hole_index = np.random.randint(0,self.geneList.shape[0])\n",
    "      \n",
    "      while 1: # If that hole is occupied then choose again until it finds an empty one \n",
    "        if not self.geneList[random_hole_index][0]:\n",
    "          break\n",
    "        else:\n",
    "          random_hole_index = np.random.randint(0,self.geneList.shape[0])\n",
    "\n",
    "      # Insert Individual in that hole\n",
    "      larvae = Individual(random_C_generator(K), random_W_generator(K,optimalD))\n",
    "      self.geneList[random_hole_index][0] = larvae\n",
    "      \n",
    "      # Once the population is set, we choose a random best gene\n",
    "      random_best_hole_index = np.random.randint(0,self.geneList.shape[0])\n",
    "      \n",
    "      while 1: # If that hole is occupied then choose again until it finds an empty one \n",
    "        if not self.geneList[random_best_hole_index][0]:\n",
    "          self.bestGene = self.geneList[random_best_hole_index][0]\n",
    "          break\n",
    "        else:\n",
    "          random_best_hole_index = np.random.randint(0,self.geneList.shape[0])\n",
    "\n",
    "    return self"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### NumPy custom functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sumMatrixes(mat1,mat2):\n",
    "  return np.add(mat1,np.asarray(mat2))\n",
    "\n",
    "def subtractMatrixes(mat1,mat2):\n",
    "  return np.subtract(mat1,np.asarray(mat2))\n",
    "\n",
    "def dotMultiply(mat1,mat2):\n",
    "  return np.multiply(mat1,np.asarray(mat2))\n",
    "\n",
    "def multiplyMatrixes(mat1,mat2):\n",
    "  return np.matmul(mat1,np.asarray(mat2))\n",
    "\n",
    "def dotDivide(mat1,mat2):\n",
    "  return np.divide(mat1,np.asarray(mat2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "def random_C_generator(K):\n",
    "  # numpy.random.randint/uniform returns from: [low,high).\n",
    "  return np.random.randint(low= 0, high= 1 + 1, size=[1,K])\n",
    "\n",
    "def random_W_generator(K,D):\n",
    "  return np.random.uniform(low= -1, high= 1, size=[K,D])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reproduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def InternalReproduction(gen):\n",
    "\n",
    "  # Mutate chromosome \n",
    "  mutate_point = np.random.randint(0, len(gen.c)) \n",
    "  \n",
    "  if(gen.c[0][mutate_point] == 0):\n",
    "    gen.c[0][mutate_point] = 1\n",
    "  elif(gen.c[0][mutate_point] == 1):\n",
    "    gen.c[0][mutate_point] = 0\n",
    "\n",
    "  # Mutate weights\n",
    "  nRows = gen.w.shape[0]\n",
    "  nColumns = gen.w.shape[1]\n",
    "  mutate_column = np.random.randint(0, nColumns) \n",
    "  \n",
    "  gen.w[:,mutate_column] = np.random.uniform(-1,1,nRows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ExternalReproduction(father,mother):\n",
    "  crossover_c_point = np.random.randint(0, len(father.c))\n",
    "  crossover_column_w_point = random.randint(0, len(father.w[0]))\n",
    "\n",
    "  # Offsprings\n",
    "  offspring1 = Individual( c = np.concatenate((father.c[:crossover_c_point], mother.c[crossover_c_point:])), \n",
    "                           w = np.concatenate((father.w[:,:crossover_column_w_point], mother.w[:,crossover_column_w_point:]), axis=1) )\n",
    "  \n",
    "  offspring2 = Individual( c = np.concatenate((mother.c[:crossover_c_point], father.c[crossover_c_point:])), \n",
    "                           w = np.concatenate((mother.w[:,:crossover_column_w_point], father.w[:,crossover_column_w_point:]), axis=1) )\n",
    "                       \n",
    "  return [offspring1,offspring2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "def RouletteWheelSelection(population): \n",
    "  # Create a population for the chosen individuals of the Roulette\n",
    "  roulettePopulation = Population(population.size)\n",
    "\n",
    "  # Make sure we don't lose our best gene in the roulette \n",
    "  roulettePopulation.insertBestGene(population.bestGene)\n",
    "\n",
    "  # Total population fitness (S)\n",
    "  S = sum([chromosome.fitness for chromosome in population.geneList])\n",
    "  \n",
    "  # Population chromosomes' relative probabilities\n",
    "  rel_prob = [chromosome.fitness/S for chromosome in population.geneList]\n",
    "\n",
    "  for i in range(population.size - 1):\n",
    "    # Generate a random uniform number (r) - (0,1]\n",
    "    r = np.random.uniform()\n",
    "    # Find the first index for which q_i < r\n",
    "    for index,individual in enumerate(population.geneList):\n",
    "      r -= rel_prob[index]\n",
    "      if r < 0:\n",
    "        roulettePopulation.addGeneToList(individual)\n",
    "        roulettePopulation.addFitnessToList(individual.fitness)\n",
    "        break\n",
    "\n",
    "  return roulettePopulation "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This functions reproduces a population.\n",
    "# It crossover the parents (external reproduction) and mutate the offspring (internal reproduction).\n",
    "# Returns the input population, updated with the crossover and mutated childs\n",
    "\n",
    "# ACK Works Great\n",
    "def ReproducePopulation(population):\n",
    "\n",
    "  childs = [] # list to store future offsprings\n",
    "  crossoverProb = np.random.uniform() # crossover probability\n",
    "  mutationProb = np.random.uniform() # mutation probability\n",
    "  geneListSize = len(population.geneList)\n",
    "\n",
    "  for index,individual in enumerate(population.geneList):\n",
    "    # Generate a random number and check if it's below the crossover probability\n",
    "    if(random.random() < crossoverProb):\n",
    "      random_mother_index = np.random.randint(0,geneListSize)\n",
    "\n",
    "      while(random_mother_index == index): # Make sure that the mother is a different individual in the population\n",
    "        random_mother_index = np.random.randint(0,geneListSize)\n",
    "\n",
    "      mother = population.geneList[random_mother_index]\n",
    "\n",
    "      # Crossover parents\n",
    "      offsprings = ExternalReproduction(individual, mother)\n",
    "\n",
    "      # Mutate offsprings\n",
    "      if(random.random() < mutationProb): \n",
    "        InternalReproduction(offsprings[0])\n",
    "\n",
    "      if(random.random() < mutationProb):\n",
    "        InternalReproduction(offsprings[1])\n",
    "\n",
    "      childs.append(offsprings[0])\n",
    "      childs.append(offsprings[1])\n",
    "  \n",
    "  # Update population's Genes list with newly-created childs\n",
    "  population.geneList = np.concatenate((population.geneList, childs))\n",
    "\n",
    "  return population"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Swarm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "def UpdateVelocitiesAndPositionsPSO(swarm, w, c1, c2):\n",
    "  for p in swarm.geneList:\n",
    "    # Velocities\n",
    "    r1 = np.random.uniform()\n",
    "    r2 = np.random.uniform()\n",
    "\n",
    "    cVt_i = (w * p.cVelocity) + (c1*r1*(p.bestFitness - p.cPosition)) + (c2*r2*(swarm.bestGene.cPosition - p.cPosition))\n",
    "    wVt_i = (w * p.wVelocity) + (c1*r1*(p.bestFitness - p.wPosition)) + (c2*r2*(swarm.bestGene.wPosition - p.wPosition))\n",
    "\n",
    "    print(\"cVt_i:\",cVt_i)\n",
    "    print(\"wVt_i:\",wVt_i)\n",
    "\n",
    "    # Positions\n",
    "    velocityProbability = 2 / math.pi * math.atan((math.pi*0.5)*cVt_i) #|2â„ðœ‹ Ã— arctan ((ðœ‹ 2) Ã— ð‘‰ð‘¡+1\n",
    "    if np.random.uniform() < velocityProbability:\n",
    "      cPt_i = 1\n",
    "    else:\n",
    "      cPt_i = 0\n",
    "\n",
    "    wPt_i = p.wPosition + wVt_i\n",
    "\n",
    "    print(\"wPt_i:\", wPt_i)\n",
    "    # -- for END --"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data Processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "CSV_NAME = \"parkinsons.csv\"\n",
    "CSV_PATH = \"E:/Perfil/OneDrive/Escritorio/MrRobot/IITV/4/TFG/data-Vito-PC/\" + CSV_NAME\n",
    "df = pd.read_csv(CSV_PATH, sep=\" \", header=None)\n",
    "\n",
    "end = df.shape[1]\n",
    "X = df.iloc[:, :end-1] # iloc[rows,[cols_start,cols_end)] <- Dataframe object\n",
    "Y = df.iloc[:, end-1] # iloc[rows,col_index] <- Dataframe series\n",
    "\n",
    "# IMPORTANT, otherwise we would be using DataFrames and matrix operations won't work\n",
    "X = X.to_numpy()\n",
    "Y = Y.to_numpy()\n",
    "J = len(np.unique(Y))\n",
    "\n",
    "N,K = X.shape[0],X.shape[1] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Scaling (min-max normalization)\n",
    "X_scaled = (X - X.min()) / (X.max() - X.min())\n",
    "\n",
    "# First partition training/test\n",
    "X_train, X_test, y_train, y_test = train_test_split(X_scaled,Y, test_size=0.33)\n",
    "\n",
    "# Second partition validation\n",
    "X_trainVal, X_testVal, y_trainVal, y_testVal = train_test_split(X_train,y_train, test_size=0.33)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ELM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Outputs the fitness for a given Gene\n",
    "def ComputeChromosomeFitness(weigths, D,C,trainingX, trainingY, testX, testY):\n",
    "    # W\n",
    "    W = weigths \n",
    "\n",
    "    # Bias\n",
    "    Bias = np.random.uniform(low= 0, high= 1, size=[D,1])\n",
    "\n",
    "    # Amplify the matrix to the size of Xtraining and Xtest\n",
    "    BiasTrainingMatrix = np.matlib.repmat(Bias,1,trainingX.shape[0])\n",
    "    BiasTestMatrix = np.matlib.repmat(Bias,1,testX.shape[0])\n",
    "\n",
    "    # Transpose BiasMatrix\n",
    "    BiasTrainingMatrix = np.transpose(BiasTrainingMatrix)\n",
    "    BiasTestMatrix = np.transpose(BiasTestMatrix)\n",
    "\n",
    "    # H (Sigmoide function) \n",
    "    activationTraining = multiplyMatrixes(trainingX, W) + BiasTrainingMatrix\n",
    "    activationTest = multiplyMatrixes(testX, W) + BiasTestMatrix\n",
    "    \n",
    "    H_Training = dotDivide( 1, (1 + np.exp(-activationTraining)) )    \n",
    "    H_Test = dotDivide( 1, (1 + np.exp(-activationTest)) )\n",
    "\n",
    "    # Beta\n",
    "    aux = multiplyMatrixes( np.transpose(H_Training) , H_Training )\n",
    "    delta = np.identity(aux.shape[0]) * 10e-3\n",
    "\n",
    "    inverse = np.linalg.inv(dotDivide(np.identity(D) , C) + delta + aux)\n",
    "\n",
    "    # Complete Formula: inv( (eye(D) ./ C) + delta + aux) * H_Training' * Ytraining;\n",
    "    Beta = multiplyMatrixes(inverse, np.transpose(H_Training))\n",
    "    Beta = multiplyMatrixes(Beta, trainingY)\n",
    "\n",
    "    # Output\n",
    "    Y_predicted = multiplyMatrixes(H_Test, Beta)\n",
    "    fitness = np.linalg.norm(Y_predicted - testY)\n",
    "\n",
    "    return fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Updates the fitnesses of a given population of Genes and also, set the new best Gene \n",
    "def ComputePopulationFitness(population, D,C,trainingX, trainingY, testX, testY):\n",
    "  # Clear previous fitnesses\n",
    "  population.fitnessList.clear()\n",
    "\n",
    "  for index,individual in enumerate(population.geneList):\n",
    "    f_i = ComputeChromosomeFitness(individual.w, D, C, trainingX, trainingY, testX, testY)\n",
    "     \n",
    "      # Set fitness\n",
    "    individual.setFitness(f_i)\n",
    "    #print(\"Best Gene Fi middle:\", population.bestGene.fitness)\n",
    "    population.addFitnessToList(f_i)\n",
    "\n",
    "      # Update Best Gene in Population\n",
    "    if(individual.fitness < population.bestGene.fitness):\n",
    "      population.bestGene = individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Updates the fitnesses of a given population of Genes and also, set the new best Gene \n",
    "def ComputeSwarmFitness(swarm, D,C,trainingX, trainingY, testX, testY):\n",
    "  # Clear previous fitnesses\n",
    "  swarm.fitnessList.clear()\n",
    "\n",
    "  for index,particle in enumerate(swarm.geneList):\n",
    "  # - - - - - - - - - - - - - - - - - -\n",
    "    f_i = ComputeChromosomeFitness(particle.wPosition, D, C, trainingX, trainingY, testX, testY)\n",
    "      \n",
    "      # Set fitness\n",
    "    particle.setFitness(f_i)\n",
    "    swarm.addFitnessToList(f_i)\n",
    "\n",
    "      # Update Pb_i for each particle and find the best gene \n",
    "    if particle.fitness < particle.bestFitness:\n",
    "      particle.bestFitness = particle.fitness\n",
    "      \n",
    "      if particle.fitness < swarm.bestGene.fitness:\n",
    "        swarm.bestGene = particle\n",
    "  # - - - - - - - - - - - - - - - - - -\n",
    "    # Update Gb of the swarm\n",
    "  swarm.setBestSwarmChromosome(swarm.bestGene.c)\n",
    "  swarm.setBestSwarmWeigths(swarm.bestGene.w)\n",
    "  print(\"Global best fitness: \", swarm.bestGene.fitness)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def TrainModelAndOutputResults(weigths, D,C,trainingX, trainingY, testX, testY):\n",
    "      # W\n",
    "    W = weigths \n",
    "\n",
    "    # Bias\n",
    "    Bias = np.random.uniform(low= 0, high= 1, size=[D,1])\n",
    "\n",
    "    # Amplify the matrix to the size of Xtraining and Xtest\n",
    "    BiasTrainingMatrix = np.matlib.repmat(Bias,1,trainingX.shape[0])\n",
    "    BiasTestMatrix = np.matlib.repmat(Bias,1,testX.shape[0])\n",
    "\n",
    "    # Transpose BiasMatrix\n",
    "    BiasTrainingMatrix = np.transpose(BiasTrainingMatrix)\n",
    "    BiasTestMatrix = np.transpose(BiasTestMatrix)\n",
    "\n",
    "    # H (Sigmoide function) \n",
    "    activationTraining = multiplyMatrixes(trainingX, W) + BiasTrainingMatrix\n",
    "    activationTest = multiplyMatrixes(testX, W) + BiasTestMatrix\n",
    "    \n",
    "    H_Training = dotDivide( 1, (1 + np.exp(-activationTraining)) )    \n",
    "    H_Test = dotDivide( 1, (1 + np.exp(-activationTest)) )\n",
    "\n",
    "    # Beta\n",
    "    aux = multiplyMatrixes( np.transpose(H_Training) , H_Training )\n",
    "    delta = np.identity(aux.shape[0]) * 10e-3\n",
    "\n",
    "    inverse = np.linalg.inv(dotDivide(np.identity(D) , C) + delta + aux)\n",
    "\n",
    "    # Complete Formula: inv( (eye(D) ./ C) + delta + aux) * H_Training' * Ytraining;\n",
    "    Beta = multiplyMatrixes(inverse, np.transpose(H_Training))\n",
    "    Beta = multiplyMatrixes(Beta, trainingY)\n",
    "\n",
    "    # Prediction\n",
    "    predictedLabels = multiplyMatrixes(H_Test, Beta)\n",
    "\n",
    "    correctPrediction = 0\n",
    "    for i in range(testY.shape[0]):\n",
    "      if np.round(predictedLabels[i]) == testY[i]:\n",
    "        correctPrediction +=1\n",
    "    \n",
    "    CCR = correctPrediction / testY.shape[0]\n",
    "    return CCR * 100"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genetic Algorithm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def GeneticAlgorithm(gaPopulation, nGenerations, optimalD, optimalC, trainX, trainY, testX, testY):\n",
    "  t = 0\n",
    "  while t < nGenerations:\n",
    "      # Reproduction\n",
    "    ReproducePopulation(gaPopulation)  \n",
    "      # Compute Fitness\n",
    "    ComputePopulationFitness(gaPopulation, optimalD, optimalC, trainX, trainY, testX, testY)\n",
    "      # Selector operator\n",
    "    gaPopulation = RouletteWheelSelection(gaPopulation)\n",
    "      # Continue iterating\n",
    "    t += 1\n",
    "  \n",
    "  return gaPopulation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Particle Swarm Optimization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def PSOAlgorithm(swarm,maxIterations, w, c1, c2, dampingW, optimalD, optimalC, trainX, trainY, testX, testY):\n",
    "  t = 0\n",
    "  while t < maxIterations:\n",
    "      # Evaluate Particles' Fitness and update Pb_i and Gb_i\n",
    "    ComputeSwarmFitness(swarm, optimalD, optimalC, trainX, trainY, testX, testY)\n",
    "\n",
    "      # Update Velocity and Position of each particle\n",
    "    UpdateVelocitiesAndPositionsPSO(swarm, w, c1, c2)\n",
    "\n",
    "      # Damping inertia\n",
    "    w = w * dampingW\n",
    "\n",
    "      # Increase step\n",
    "    t += 1\n",
    "\n",
    "  return 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coral Reef Optimization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CROAlgorithm():\n",
    "  t = 0\n",
    "  return 1\n",
    "  '''\n",
    "  while t < coralGenerations:\n",
    "    # Update variables\n",
    "    F_broadcast = np.random.uniform() # broadcast fraction\n",
    "    F_brooding = 1 - F_broadcast # brooding fraction\n",
    "    Pd = 0.2 # predation probability\n",
    "    \n",
    "    # Broadcast and Brooding\n",
    "\n",
    "    # Asexual reproduction\n",
    "\n",
    "    # Predation \n",
    "\n",
    "    # Fitness Evaluation\n",
    "\n",
    "    # Increase step\n",
    "    t += 1\n",
    "    # - end WHILE\n",
    "\n",
    "  print(\"Best CRO: \", \"coral X\")\n",
    "  '''"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ea8238db6dc1fe8cf6fe83219457bde9bcbcde2053d40c59d2e78211d10c5fee"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
